crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs

```diff
@@ -0,0 +1,409 @@
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<CssRuleAction> {
+        if !state.can_be_sorted || state.first_out_of_order_pair.is_none() {
+            return None;
+        }
+
+        let mut mutation = ctx.root().begin();
+
+        let original_items = to_block_items_with_position(&state.block);
+        let mut sorted_items = original_items.clone();
+        sorted_items.sort_by(|a, b| sort_info(&a.node).cmp(&sort_info(&b.node)));
+
+        let pairs = original_items.iter().zip(sorted_items.iter());
+
+        for (item_to_replace, replacement) in pairs {
+            if item_to_replace.position != replacement.position {
+                mutation.replace_node_discard_trivia(
+                    item_to_replace.node.clone(),
+                    replacement.node.clone(),
+                );
+            }
+        }
+
+        return Some(RuleAction::<CssLanguage>::new(
+            ActionCategory::QuickFix,
+            ctx.metadata().applicability(),
+            markup! { "Sort these properties" }.to_owned(),
+            mutation,
+        ));
+    }
+}
+
+#[derive(Clone)]
+struct NodeWithPosition {
+    position: usize,
+    node: AnyCssDeclarationOrRule,
+}
+
+fn to_block_items_with_position(block: &CssDeclarationOrRuleBlock) -> Vec<NodeWithPosition> {
+    block
+        .items()
+        .into_iter()
+        .enumerate()
+        .map(|(position, item)| NodeWithPosition {
+            position,
+            node: item,
+        })
+        .collect()
+}
+
+// takes the unsorted list and returns the first item that is out of order, and the item it should be placed before
+// returns None if the list is sorted
+fn get_first_out_of_order_pair(
+    items: &[NodeWithPosition],
+) -> Option<(NodeWithPosition, NodeWithPosition)> {
+    if items.is_empty() {
+        return None;
+    }
+
+    let mut first_out_of_order_node: Option<NodeWithPosition> = None;
+
+    // find the first node that's 'less than' the previous node
+    let mut prev_item_info = sort_info(&items[0].node);
+    for item in items.iter() {
+        let item_info = sort_info(&item.node);
+        if item_info.cmp(&prev_item_info) == Ordering::Less {
+            first_out_of_order_node = Some(item.clone());
+            break;
+        }
+        prev_item_info = item_info;
+    }
+
+    // find where that node belongs
+    if let Some(first_out_of_order_node) = first_out_of_order_node {
+        let first_out_of_order_node_info = sort_info(&first_out_of_order_node.node);
+        for item in items.iter() {
+            let item_info = sort_info(&item.node);
+            if item_info.cmp(&first_out_of_order_node_info) == Ordering::Greater {
+                return Some((first_out_of_order_node.clone(), item.clone()));
+            }
+        }
+    }
+
+    None
+}
+
+// returns (vendor prefix, name without vendor prefix) in lowercase for a CssIdentifier
+fn css_identifier_to_prop_name(ident: &CssIdentifier) -> Option<(String, String)> {
```

Comment:
Let's not allocate `String` if we don't use them in diagnostics/messages.

Considering how you use this information, it's best to return a `&str` or a `&'static str`